middle_end/flambda2/types/meet_and_join.ml

```diff
@@ -167,36 +293,23 @@ and meet0 env (t1 : TG.t) (t2 : TG.t) : TG.t * TEE.t =
         ~known_canonical_simple_at_in_types_mode:simple2
     in
     match simple2 with
-    | None -> (
-      match meet_expanded_head env expanded1 expanded2 with
-      | Left_head_unchanged -> t1, TEE.empty
-      | Right_head_unchanged -> t2, TEE.empty
-      | New_head (expanded, env_extension) -> ET.to_type expanded, env_extension
-      )
+    | None ->
+      map_result ~f:ET.to_type (meet_expanded_head env expanded1 expanded2)
     | Some simple2 -> (
       match meet_expanded_head env expanded1 expanded2 with
-      | Left_head_unchanged ->
+      | Ok (Left_input, env_extension) ->
         let env_extension =
-          add_equation simple2 (ET.to_type expanded1) TEE.empty
-        in
-        TG.alias_type_of kind simple2, env_extension
-      | Right_head_unchanged -> TG.alias_type_of kind simple2, TEE.empty
-      | New_head (expanded, env_extension) ->
-        (* It makes things easier (to check if the result of [meet] was bottom)
-           to not return "=simple" in the bottom case. This is ok because no
-           constraint is being dropped; the type cannot be refined any further.
-           Same below. (Note that any bottom equation added to [env_extension]
-           for [simple2] is of no consequence since the extension will be
-           ignored; see the [meet] function below.) *)
-        let ty =
-          if ET.is_bottom expanded
-          then MTC.bottom kind
-          else TG.alias_type_of kind simple2
+          add_equation simple2 (ET.to_type expanded1) env_extension
         in
+        Ok (Right_input, env_extension)
```

Comment:
I think a comment would be good there, returning `Right_input` while we are matching on `Left_input` is confusing.

Reply:
It might also be clearer to factorize the pattern-macthing to return only the `env_extension`, and then return `Ok (Right_input, env_extension)` outside the match.